Koreader progress sync interface

MFDeAngelo opened the pull request:
# Added
- Added: Added native support to sync progress with Koreader. See [this issue](https://github.com/koreader/koreader/issues/11105) on the Koreader repository for more information.

MFDeAngelo commented:
This PR is not ready for a merge (if that isn't already apparent.) I wanted to show you the basic interface that I'd like to implement and get some feedback on it before diving in. Here are some relevant details that will help you understand the scope of the proposed change.

As it currently stands, you can read books on Koreader by using OPDS and stream the book from Kavita. Unfortunately, I have two issues with this. First, I haven't even gotten it to work yet. I've only tried a handful of times, but it always produces just the Koreader logo. It might simply be user error, but if you like, I can investigate further and open an issue if necessary. Second, what if I'm away from my LAN and want to sync my books? What if the power goes out? Well too bad. This feature should solve that issue.

Koreader has a plugin that ships with the install called 'progress sync'. It is on by default. It allows a user to connect to a Koreader sync server, to sync progress across Koreader devices. We can use this functionality within Koreader to sync progress between Koreader devices and Kavita. A user would only need to us OPDS to download the book onto their Koreader device, then authenticate to the server in the progress sync page, and poof! The Koreader device and Kavita can sync progress!

The user will have to enter the following url into their Koreader device: 'http://my-kavita-server.com/api/koreader/api-key/'. Then they will login using their username, and the password can be anything, as the authentication happens through the api-key. (I don't think it makes sense to use the user's actual password, as Koreader sends the MD5 hash of the password to the server. Since Kavita doesn't store passwords as an MD5, it would make implementing this much more difficult.) Alternatively we could require the user to use the api-key as the password. Your advice would be appreciated.

Koreader allows you to register as a user. I think it's in our best interest to leave that functionality out.

There are Put and Get endpoints that allow for syncing between the device and the server. Koreader uses the MD5 hash of the book (or filename as a secondary option) to identify what book is being synced. I'm not sure how to handle the device id stuff yet, but that seems like an implementation concern.

Let me know what you think. I'd love to implement this functionality.
Also, I'd like to thank you for putting together this wonderful piece of software. I've really enjoyed using it. It was the second thing I installed on my home server, and I'm glad I did!

majora2007 commented:
First off, I would welcome a PR for this (and any needed PR for OPDS and Koreader). 

> (I don't think it makes sense to use the user's actual password, as Koreader sends the MD5 hash of the password to the server. Since Kavita doesn't store passwords as an MD5, it would make implementing this much more difficult.) Alternatively we could require the user to use the api-key as the password. Your advice would be appreciated.

I agree. I don't want to use passwords with external applications, hence why we have the api key in the first place. 

> Koreader allows you to register as a user. I think it's in our best interest to leave that functionality out.

Agreed. External user provisioning is not something I'm a fan of for Kavita. 


My main points of concern are this: 
1. Make sure you use Dto classes when appropriate with good comments
2. Try to write some unit tests on the actual sync (unless it's as simple as calling SaveProgress())
3. Let's make sure the code is separated into a Koreader controller (which you've already done) and any Koreader service if needed.
4. Please have in mind that documentation will be needed for the wiki as well once this is implemented. 

Let me know if you need anything either via Github or via discord.

MFDeAngelo commented:
Wonderful! I'll be working on this over the next couple of weeks. I've looked through code, and I think I have a general idea of how to accomplish it. Thanks for your help!

MFDeAngelo committed:
Basics structure of koreader support including db migration

MFDeAngelo commented:
Not sure why this says it was closed by me. I had no intention of closing it yet.
Just wanted to give an update saying that I'm still working on it. I've been able to establish a connection from koreader to kavita, and now am working on converting the progress datatypes from one to another.

majora2007 commented:
Awesome. I think AutoMapper can help with that as much as possible. It's okay if the PR gets closed, as long as you re-open once you're ready for some more review.

MFDeAngelo committed:
koreader base complete

MFDeAngelo commented:
@majora2007 I've made some more changes and have it working locally. When reading in koreader, the progress is updated on kavita, and vice versa. However, I don't feel confident that I've accounted for every potential "bookScrollId". I wanted to know what you thought about releasing this feature in beta to start off with, with the expectation that it might break here and there.

I'll get this branch caught up on the develop branch in the next couple of days, improve the logging, and start preparing the Wiki entry. Could you perform a quick once over to give me any feedback on it?

majora2007 commented:
Yeah, ill go over it and give you a review. I think we can probably get this into the current nightly, since we still have some good amount of time before the next stable goes out.

MFDeAngelo committed:
Fixed koreader tests

majora2007 commented on API/API.csproj:
```diff
@@ -12,9 +12,6 @@
     <LangVersion>latestmajor</LangVersion>
   </PropertyGroup>
 
-  <Target Name="PostBuild" AfterTargets="Build" Condition=" '$(Configuration)' == 'Debug' ">
```
You're going to want to commit this back in.

MFDeAngelo replied:
Oops! Didn't mean to commit this! Adding it back now.

majora2007 commented on API/Controllers/KoreaderController.cs:
```diff
@@ -0,0 +1,94 @@
+    {
+        var userId = await GetUserId(apiKey);
+        var user = await _unitOfWork.UserRepository.GetUserByIdAsync(userId);
+        return Ok(new { username = user.UserName });
```
Please have newlines above any return statements. 

For the sake of a full review, please use Dto Objects under a DTOs/Koreader folder, so we can easily track and understand the interfaces. 

Also add documentation on the APIs

MFDeAngelo replied:
I have a DTO in the DTOs/Koreader folder. I'm not sure what other DTO's you'd like me to add there. Are you suggesting that I change the responses from anonymous types to concrete DTOs?

majora2007 replied:
Exactly that.

majora2007 commented on API/Controllers/KoreaderController.cs:
```diff
@@ -0,0 +1,94 @@
+        var response = new
+        {
+            document = request.Document,
+            timestamp = DateTime.Now
```
I feel like this should be UtcNow. Otherwise depending on where the server is, Koreader will have different account for timestamp.

majora2007 commented on API/Controllers/KoreaderController.cs:
```diff
@@ -0,0 +1,94 @@
+    {
+        var userId = await GetUserId(apiKey);
+        await _koreaderService.SaveProgress(request, userId);
+        var response = new
```
You can inline this response

majora2007 commented on API/Controllers/KoreaderController.cs:
```diff
@@ -0,0 +1,94 @@
+    {
+        var userId = await GetUserId(apiKey);
+        var response = await _koreaderService.GetProgress(ebookHash, userId);
+        Console.WriteLine(response.Progress);
```
Use a logger. Might want to put this on Trace so if we need to debug, we can.

majora2007 commented on API/DTOs/Koreader/EpubPositionDto.cs:
```diff
@@ -0,0 +1,8 @@
+
+public class EpubPositionDto
+{
+    public string KavitaPosition { get; set; }
```
I'd love some extra documentation to explain what these fields are. Because once it's merged, I will own it.

MFDeAngelo replied:
I ended up removing this. It felt weird to pass this object around, as it was often immediately remapped to another.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+    public static string HashContents(string filePath)
+    {
+
+        if (string.IsNullOrEmpty(filePath))
```
Move return on same line as if or put {} around

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+        if (string.IsNullOrEmpty(filePath))
+            return null;
+
+        using (var file = File.OpenRead(filePath))
```
We don't need scoped using aka using (). You can just put using on the same line and let .net do the heavy work for us.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+        if (string.IsNullOrEmpty(filePath))
+            return null;
+
+        using (var file = File.OpenRead(filePath))
```
We need to use _directoryService.FileSystem.File. This allows us to mock the filesystem.

MFDeAngelo replied:
This presents an interesting problem. 
KoreaderHelper is currently a static class. In order to add DirectoryService as a member of KoreaderHelper, I have to make it a concrete type. If KoreaderHelper is a concrete type, then it needs to be injected into MangaFileBuilder using dependency injection. MangaFileBuilder cannot be instantiated in the dependency tree (or composition root. Whatever you call it) since it is supposed to be instantiated ad hoc.
I can add a static member to KoreaderHelper and insert a DirectoryService object in, but that feels wrong. Do you have an alternative idea for how to achieve the mocking capability?

majora2007 replied:
Hmm, it all comes down to the unit tests. I needed to use a mocked filesystem for a lot of the unit tests to emulate last write times and whatnot, but technically, if we have some non-copywrited files in TestData, we can avoid this whole mess. 

I just want to make sure we have good test coverage on KoreaderHelper.

MFDeAngelo replied:
Understandably. I think the underlying issue is that we are trying to mock out static functionality (reading a file) using a concrete object.

I haven't thought this through all the way, but what if we added the mocking capability to a static object? It obviously has its drawbacks...
```
public static class KavitaFile
{
    private Dictionary<string, Stream> mockedFiles;

    public static Stream OpenRead(string path)
    {
        if (mockedFiles.Any())
        {
            return File.OpenRead(path);
        }
        else
        {
            return mockedFiles[path];
        }
    }

    public static void AddMockedFile(string path, Stream stream)
    {
        mockedFiles.Add(path, stream);
    }
}
```
This is incomplete but gets the idea across. Before any tests run, you need to set up the predefined mock files. It makes sense to me that you'd only want one mocking strategy, so I'm assuming you'd need to convert all places that you're using the directory service.
The most obvious drawback is abusing the mock functionality in application code. I don't think I'd include this in the current PR. Is this a route worth pursuing?

Otherwise, I can grab a few books of Project Gutenberg for test data.

majora2007 replied:
I think let's just grab some files. Moving to Manga files adds a lot of mess and really not needed. As long as we can test, I'm happy.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+            file.Close();
+            md5.TransformFinalBlock(new byte[0], 0, 0);
+            byte[] hashBytes = md5.Hash;
+            return BitConverter.ToString(hashBytes).Replace("-", "").ToUpper();
```
Use string.Empty

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+
+            file.Close();
+            md5.TransformFinalBlock(new byte[0], 0, 0);
+            byte[] hashBytes = md5.Hash;
```
Remove the hashBytes variable, it's not needed.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+        else
+        {
+            progress.BookScrollId = $"//html[1]/BODY/APP-ROOT[1]/DIV[1]/DIV[1]/DIV[1]/APP-BOOK-READER[1]/DIV[1]/DIV[2]/DIV[1]/DIV[1]/DIV[1]/{lastTag}";
```
Help me understand what this is

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+    public static void UpdateProgressDto(string koreaderPosition, ProgressDto progress)
+    {
+        var path = koreaderPosition.Split('/');
+        var docNumber = path[2].Replace("DocFragment[", "").Replace("]", "");
```
Please add bounds checking so we don't throw when something isn't as expected.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+        var path = koreaderPosition.Split('/');
+        var docNumber = path[2].Replace("DocFragment[", "").Replace("]", "");
+        progress.PageNum = Int32.Parse(docNumber) - 1;
+        string lastTag = path[5].ToUpper();
```
Don't use explicit types, use var whenever possible.

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,94 @@
+    {
+        string lastTag;
+        int koreaderPageNumber = progressDto.PageNum + 1;
+        if (string.IsNullOrEmpty(progressDto.BookScrollId))
```
if statements always need {} unless a simple jump (return/break/continue) that can be on one line.

majora2007 commented on API/Services/KoreaderService.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        var file = await _unitOfWork.MangaFileRepository.GetByHash(koreaderBookDto.Document);
+        var userProgressDto = await _unitOfWork.AppUserProgressRepository.GetUserProgressDtoAsync(file.ChapterId, userId);
+        KoreaderHelper.UpdateProgressDto(koreaderBookDto.Progress, userProgressDto);
```
You need to break up the code into sections so it's easier to understand. The first 2 lines are variable setup. Next 2 are the actual logic. Last is the commit, so break into 3 sections with a newline inbetween.

majora2007 commented on API/Services/KoreaderService.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            Document = bookHash,
+            Device_id = "kavita",
+            Device = "kavita",
```
Kavita should be capital. Is device_id supposed to be a unique identifier? Could we use the ServerId? That is unique unless someone changes their token in appsettings.json (which usually only happens on reinstalling incorrectly).

MFDeAngelo replied:
ServerId would make more sense if it is a unique identifier. Could you tell me where to get that from? I can't find anything by searching `ServerId` or its synonyms.

majora2007 replied:
You should find this on BuildInfo I believe, but maybe getting from the db is better. You can get from setting item. I believe it also goes as InstallId.

majora2007 commented on API/Services/KoreaderService.cs:
```diff
@@ -0,0 +1,52 @@
+            Device_id = "kavita",
+            Device = "kavita",
+            Progress = koreaderProgress,
+            Percentage = 0.5f
```
Curious why we are sending a percentage if it's not correct...

MFDeAngelo replied:
By what I've seen, Koreader doesn't appear to use this value. I think it could be used by the Koreader progress sync server to display statistics or something. I'll see if I can track it down a little better.

MFDeAngelo replied:
So I've been looking at this even more, and [it appears to be only for message boxes.](
https://github.com/koreader/koreader/blob/master/plugins/kosync.koplugin/main.lua#L798)
A [little earlier in the method](https://github.com/koreader/koreader/blob/master/plugins/kosync.koplugin/main.lua#L764) it compares against a previous percentage and a new percentage to see if if is in the same position. If so, it isn't supposed to update the progress in Koreader. However, in testing it doesn't behave that way. Maybe it has to do with direct comparison of a floating point.

majora2007 replied:
It looks like it's a user friendly percentage. Can we not use the Progress mapped to a 0-1 float?

majora2007 replied:
Someone from Stump did some research on this:
https://github.com/stumpapp/stump/issues/239#issuecomment-2428256328

majora2007 commented on API/Services/Tasks/Scanner/ProcessSeries.cs:
```diff
@@ -818,6 +818,7 @@ private void AddOrUpdateFileForChapter(Chapter chapter, ParserInfo info, bool fo
             var file = new MangaFileBuilder(info.FullFilePath, info.Format, _readingItemService.GetNumberOfPages(info.FullFilePath, info.Format))
                 .WithExtension(fileInfo.Extension)
                 .WithBytes(fileInfo.Length)
+                .WithHash()
```
I'd love to know the overhead of hashing files, perhaps a benchmark can help us understand. (See the Benchmark project)

MFDeAngelo replied:
I set up the benchmark and will be committing it soon. I set up the benchmark to build a MangaFile using the MangaFileBuilder. I went ahead and ran the benchmark on hashing a 300 KB epub. Here are the results:
```
| Method                  | Mean      | Error     | StdDev    | Median    | Ratio | RatioSD | Rank | Allocated | Alloc Ratio |
|------------------------ |----------:|----------:|----------:|----------:|------:|--------:|-----:|----------:|------------:|
| TestBuildManga_baseline |  40.18 us |  1.631 us |  4.731 us |  38.49 us |  1.00 |    0.00 |    1 |     922 B |        1.00 |
| TestBuildManga_withHash | 230.09 us | 10.654 us | 31.415 us | 229.86 us |  5.81 |    0.99 |    2 |    7151 B |        7.76 |
```

MFDeAngelo replied:
One thing to keep in mind about the hashing algorithm is that a 300KB file will only have 4KB of its contents hashed. At most, the algorithm will hash 10 KB of the file. That means that even though there are files much larger than 300KB, they shouldn't take that much longer to hash.

majora2007 replied:
For Koreader, is this only for epub files? Because if so, I think we could swallow this type of overhead. If it's for archives as well, then we might want to look into offloading this into a separate task that performs the hashing. This will offload the extra spend on the scan loop and shortly make it available after scan for syncing.

MFDeAngelo replied:
I switched it to only hash EPUBs. I don't think there is any sense in hashing the other files as I don't believe Koreader supports Archives. It supports PDFs though. We can add support for that later.

majora2007 commented on API/Services/Tasks/Scanner/ProcessSeries.cs:
```diff
@@ -590,8 +590,8 @@ private void DeterminePublicationStatus(Series series, List<Chapter> chapters)
 
-            var maxVolume = (int) (nonSpecialVolumes.Any() ? nonSpecialVolumes.Max(v => v.MaxNumber) : 0);
-            var maxChapter = (int) chapters.Max(c => c.MaxNumber);
+            var maxVolume = (int)(nonSpecialVolumes.Any() ? nonSpecialVolumes.Max(v => v.MaxNumber) : 0);
```
Please add the space between the cast back

majora2007 reviewed (changes requested):
Lots of style comments and the need for documentation to get the full gist. 

I def would recommend adding some links to places where you learned to reverse engineer, which will not only help me, but any other project that wants to add this sort of support.

majora2007 commented on API/Data/Migrations/20240704154945_BookHash.cs:
```diff
@@ -0,0 +1,28 @@
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<string>(
+                name: "Hash",
```
Second thought, let's make this KoreaderHash given they use a custom MD5 hashing algorithm.

MFDeAngelo committed:
Renamed Hash to KoreaderHash and added Koreader documentation

MFDeAngelo committed:
Merge branch 'develop' into develop

MFDeAngelo commented:
@majora2007, I've merged the most recent changes into my branch and updated the PR. I've made a few small changes since the last time I ran the project. however, I noticed that I can't run the project anymore. I launch the API project and it directs me to localhost, but it returns an HTTP 500.30. I even pulled straight from the Kavita develop branch, and am getting the same result. I'm not sure if it is my local setup, or if there is something wrong in develop. 

Regardless, I'd like to run a few quick tests, then I'm ready for an in depth review. If we can resolve the HTTP 500 error, then I'll run those tests and get it to you to review.

MFDeAngelo committed:
MangaFiles can only have a Koreader hash if it is an EPUB.

MFDeAngelo committed:
Updated migrations for koreader hash

majora2007 commented:
I just pulled down your code and the webapp is working just fine, I'm not able to reproduce any 500 errors. Are you also launching the webui server as well with angular? You should be going to localhost:4200/

majora2007 commented on API/Entities/MangaFile.cs:
```diff
@@ -21,6 +21,10 @@ public class MangaFile : IEntityDate
     /// <summary>
+    /// A hash of the document using Koreader's unique hashing algorithm
+    /// </summary>
+    public string KoreaderHash { get; set; }
```
This should be `string?` as we aren't using for all file types. I would also update `<remarks>` to include information about the supported file types.

majora2007 commented on API/Helpers/Builders/MangaFileBuilder.cs:
```diff
@@ -60,4 +60,13 @@ public MangaFileBuilder WithId(int id)
+        {
+            _mangaFile.KoreaderHash = KoreaderHelper.HashContents(_mangaFile.FilePath);
+        }
+        return this;
```
Newlines above returns

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,107 @@
+    /// <param name="filePath">The path to the file to hash</param>
+    public static string HashContents(string filePath)
+    {
+
```
Remove empty line

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,107 @@
+
+        using var file = File.OpenRead(filePath);
+
+        int step = 1024;
```
Use var whenever possible

majora2007 commented on API/Helpers/KoreaderHelper.cs:
```diff
@@ -0,0 +1,107 @@
+        var fileName = Path.GetFileName(filePath);
+        var fileNameBytes = Encoding.ASCII.GetBytes(fileName);
+        var bytes = MD5.HashData(fileNameBytes);
+        return BitConverter.ToString(bytes).Replace("-", string.Empty);
```
Newline above returns

majora2007 reviewed (commented):
I did a code review, I will pull this down soon and start testing the heck out of it. 

@DieselTech if you want to also start testing. 

We missed our opportunity for the v0.8.3 release, but v0.8.4 development is right around the corner and I know many people will want this feature.

MFDeAngelo committed:
Koreader stylistic changes

MFDeAngelo commented:
@majora2007 I fixed the stylistic concerns you had. I'm not sure how exactly to map a progress to a float though. It seems like it could be difficult, because of the way Kavita tracks a user's progress. I could use currentPage/totalPages, but that isn't the same way that Koreader tracks progress, and would lead to discrepancies. Do you have any other ideas?

majora2007 commented:
> @majora2007 I fixed the stylistic concerns you had. I'm not sure how exactly to map a progress to a float though. It seems like it could be difficult, because of the way Kavita tracks a user's progress. I could use currentPage/totalPages, but that isn't the same way that Koreader tracks progress, and would lead to discrepancies. Do you have any other ideas?

Well Kavita does technically use currentPage/totalPage. It just stores an additional xpath to resume scroll position, but the underlying progress would be currentPage/totalPage. This would ensure that the modal on Koreader side shows something correct. 

I'll try and pull this down this weekend and give it a round of review. I don't have Koreader, but we can deploy this to a test instance and get some testers before we merge it into develop.

majora2007 commented:
@DieselTech Can we get this deployed to the dev instance so we can ask users to test it? I don't want to merge into nightly yet without more validation testing from people with actual devices. We can lock the users down to dedicated testers.

majora2007 commented:
Just to update you, I'm planning to put this in a canary and asking for some testing before I merge it into develop. Canary is currently being used for a large refactor of the scan loop, to ensure it's not breaking anything before moving the code into nightly/develop. Once that wraps, this is up next.

calilex commented:
This feature is great, I’m looking forward to it very much.

majora2007 commented:
Just an update, canary should be coming to an end today/tomorrow. It's been nearly a week of just working on rewriting the scanner but we are at a good merge point. Once that occurs, I will be creating a new canary with this for testing (which should be quick, I already have a user to perform the extended testing). 

Thank you for your patience. With bigger PRs, it's critical to ensure things work well before merging in and taking ownership.

majora2007 commented:
If you could resolve the conflicts then I'll get you merged into canary and start the testing. 

Discussion for testing: 
https://discord.com/channels/821879810934439936/1298773623825367170

majora2007 committed:
Lots of Bugfixes (#3308)

majora2007 committed:
Bump versions by dotnet-bump-version.

majora2007 committed:
Update to Angular 18 (#3309)

majora2007 committed:
Bump versions by dotnet-bump-version.

MFDeAngelo commented:
Hi @majora2007, apologies for the delay. I've never worked with synchronizing a fork. I finally understand what having multiple remotes is for now, so I think I'll be a little better at merging these things.

Anyways, I've brought the branch up to date with the feature/koreader-sync branch.

majora2007 commented on API/Entities/MangaFile.cs:
```diff
@@ -21,6 +21,11 @@ public class MangaFile : IEntityDate
     public required string FilePath { get; set; }
     /// <summary>
+    /// A hash of the document using Koreader's unique hashing algorithm
+    /// <remark> KoreaderHash is only available for epub types </remark>
```
Reminder to self, I need to fix this (remarks should be outside of summary tag)

majora2007 commented on API/Helpers/Builders/MangaFileBuilder.cs:
```diff
@@ -60,4 +60,14 @@ public MangaFileBuilder WithId(int id)
+
+    public MangaFileBuilder WithHash()
+    {
+        if (_mangaFile.Format == MangaFormat.Epub)
```
Reminder to self, inversion of control this (aka check if not epub and return this)

MFDeAngelo committed:
Merged before Koreader push

MFDeAngelo committed:
Merge changes from only including epubs in koreader update

MFDeAngelo committed:
sync to koreader feature branch

MFDeAngelo committed:
Add logging to koreader service and return Kavita's percentage to koreader progress sync

majora2007 commented:
I'm going to do some cleanup for you and once done, I encourage you to read over the upcoming PR into canary. 

Some items I need from you:
* Descriptions for the KoreaderController methods
* Acknowledgement the epub included is not under copywrite

majora2007 commented:
@MFDeAngelo A few more things that I'll need implementation help on. In the KoreaderService, there are a lot of cases of data coming back null that aren't being handled. I'm going to augment the code with TODOs and push it up. Please branch off my branch and get everything in order. 

You wont have to deal with any merges anymore since the code is in my own branch. I just need help since I don't have Koreader setup. 

You can see the TODOs on the PR:
https://github.com/Kareadita/Kavita/pull/3311

john-okeefe commented:
I'm very happy to see this feature being worked on. I do hope it can be implemented soon.
